sql/analyzer/apply_foreign_keys.go

```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
 	}
 }
 
+// getForeignKeyEditor merges both getForeignKeyReferences and getForeignKeyRefActions and returns a single editor.
+func getForeignKeyEditor(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	fkEditor, err := getForeignKeyReferences(ctx, a, tbl, cache, fkChain)
+	if err != nil {
+		return nil, err
+	}
+	fkEditor2, err := getForeignKeyRefActions(ctx, a, tbl, cache, fkChain)
+	if err != nil {
+		return nil, err
+	}
+	// Due to tables not containing their database, we have to grab the table's database from the foreign key definition.
+	// This is so we may make proper use of cached updaters. Therefore, both calls return a full editor rather than just
+	// their respective specialties (and may be used independently).
+	if fkEditor == nil && fkEditor2 == nil {
+		return nil, nil
+	} else if fkEditor != nil && fkEditor2 != nil {
+		fkEditor.RefActions = fkEditor2.RefActions
+		return fkEditor, nil
+	} else if fkEditor != nil {
+		return fkEditor, nil
+	} else {
+		return fkEditor2, nil
+	}
+}
+
+// getForeignKeyReferences returns an editor containing only the references for the given table.
+func getForeignKeyReferences(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	var editor sql.ForeignKeyUpdater
+	declaredFks, err := tbl.GetDeclaredForeignKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// We can ignore foreign keys that have been previously used as we know that the data we're referencing is being inserted
+	for i := len(declaredFks) - 1; i >= 0; i-- {
+		if fkChain.HasForeignKey(declaredFks[i].Name) {
+			newDeclaredFks := make([]sql.ForeignKeyConstraint, 0, len(declaredFks)-1)
+			newDeclaredFks = append(newDeclaredFks, declaredFks[:i]...)
+			newDeclaredFks = append(newDeclaredFks, declaredFks[i+1:]...)
+			declaredFks = newDeclaredFks
+		}
+	}
```

Comment:
```suggestion
	newDeclaredFks := make([]sql.ForeignKeyConstraint, 0, len(declaredFks))
	for _, fk := range declaredFks {
		if !fkChain.HasForeignKey(fk.Name) {
			newDeclaredFks = append(newDeclaredFks, fk)
		}
	}
	declaredFks = newDeclaredFks
```

Reply:
This is why PR reviews are good, as that is some horrendous code I wrote 🤣